documentation changes March 2025 (#466)
EnergySystemsModellingLab/MUSE_2.0: docs/model_description.md

2) [Commodity Price Discovery](#2-commodity-price-discovery)
3) [Agent Investment](#3-agent-investment)
4) [Carbon Budget Solution (or CO<sub>2</sub> Price Responsiveness)](#4-carbon-budget-solution-or-co2-price-responsiveness)
5) [Recursively Solve Using Steps (2) to (5) for Each Milestone Year until End](#5-recursively-solve-using-steps-2-to-5-for-each-milestone-year-until-end)

## Framework Processing Flow

The MUSE 2.0 solution concept is as follows:

### 1. Initialisation

Read input data, performing basic temporal set up, commodity and process/asset information.
A range of input data consistency checks are performed.

### 2. Commodity Price Discovery

> Dispatch Optimimisation (hereon "dispatch") is executed to determine commodity production

Reviewer: ```suggestion
Dispatch Optimisation (hereon "dispatch") is executed to determine commodity production
```

and consumption, with fixed asset capacities. In the first milestone year - the calibrated
base year - this step is performed before any agent investment step. In later
milestone years, it is performed after agent investment is complete to determine prices for the
following milestone year alongside production/consumption of all commodities.

#### A. Price discovery via asset dispatch optimisation using dual solution

Asset dispatch is determined via linear programming, where the objective function is the cost of
operating the system over a year, which must be minimised. The dual solution of this optimisation
are used to discover commodity price. The methodology for dispatch is described in detail in the
**Dispatch Optimisation** section of this documentation.

**Price** is discovered for each commodity, for each time slice, and for each region, using the
dual solution from the dispatch optimisation. Price is determined for each time slice by

> (a) taking the dual value for the commodity balance constraint, (b) subtracting the dual value

Author: Yes!

of the capacity/availability constraint (this step is performed separately for each asset/process),
and then (c) taking the maximum value from the set of results from steps a-b.

The result is a time sliced price for each commodity in each region.

#### B. Price discovery for commodities not produced

We also calculate the prices of commodities that are not present in the dispatch
solution, but could exist in the solution for the next milestone year. These are calculated
directly from input data. This is done by calculating the marginal price of the process producing
the commodity in question with the best decision value, where decision values are calculated
using the utilisation of the next most expensive (marginal cost) asset in the dispatch stack,

> adjusted for availability differences, and commodity prices from the price discovery at step 2(A).

Reviewer: It would also be useful to have a formula for decision values. It's not clear how utilisation, availability and commodity price should be combined here.

Reviewer: Just spoke to @tsmbland and he said decision values should be the same as objective values for the single-objective case? If so, we don't have to worry about it until we add support for additional decision rules.

Author: Actually I think this description is wrong, because if the commodity is present in the solution then we already have a price for it. Instead I think we should use an assumption of max utilisation for now, and create an issue for this (because it would be better to use the utilisation corresponding to the availability profile of the potential downstream demand - but this is a very complicated thing to do!)

Reviewer: > Also it may be possible to get prices for commodities not present in the solution by adding the processes to the problem with capacity constrained to zero. Then Highs might generate related prices - we would need to try this.

We can certainly give this a go!

Does the below make sense (for now)? I can open an issue about getting a better value for utilisation here.

```suggestion
We also calculate the prices of commodities that are not present in the dispatch
solution, but could exist in the solution for the next milestone year. These are calculated
directly from input data. This is done by calculating the marginal price of the process producing
the commodity in question with the best decision value, assuming maximum utilisation.
```

Author: OK for now I think. Can refine if this solution works.

  > **Issue 1:** There may be a better way to do this step 2(B).

>   One could add all processes that could potetentially exist in the

Reviewer: ```suggestion
  One could add all processes that could potentially exist in the
```

  milestone year to the dispatch optimisation formulation, even
  those that do not have a related asset (i.e. they have no
  capacity in the milestone year). A commodity balance constraint
  could then be created for every commodity, even those that are
  not yet produced. Marginal prices could be observed from the
  dispatch optimisation result as per other commodity prices.
  This would need to be tried out to see if it works.

### 3. Agent Investment

The capacity investment of new assets required in the next milestone year are calculated as follows:

#### A. End-of-life capacity decommissioning

  We decommission assets that have reached the end of their life in the milestone year.

#### B. Agent investment

  Starting with each service demand (SVD) commodity, for each agent that produces that commodity:

##### i. Determine potential utilisation at time slice level

  For each asset/process that produces the commodity, we calculate the potential dispatch
  in each time slice. This is done by observing the amount of demand that is above the

>   marginal cost of this process in the solution for the previous milestone year (i.e. demand

Author: Marginal cost is operational costs only (i.e. no contribution from capex and fixed opex). So it's cost of inputs, cost (if any) of outputs, and variable operating cost.

>   marginal cost of this process in the solution for the previous milestone year (i.e. demand

Reviewer: Just to double-check: the costs of inputs/outputs are commodity costs (as in, the things in `commodity_costs.csv`) rather than commodity prices... right?

>   marginal cost of this process in the solution for the previous milestone year (i.e. demand

Author: Actually we need a bit of nuace here. 
Tom is correct that we need to include both price_of_commodity and cost_of_commodity in marginal price determination. BUT we exclude price_of commodity for the commodity of interest that we are calculting the price for (this is the value we're after) - only include this for other commodity inputs/outputs.

N.B. I think price can actually be negative
N.B. Where there are multiple PACs, we include the price_of_commodity for other PACs. This is debateable however!
N.B. I feel like the parameter name cost_of_commodity is confusing. Can we change it to commodity_levy? Then document that a levy can be negative?

>   marginal cost of this process in the solution for the previous milestone year (i.e. demand

Reviewer: All interesting info! But if we're happy with the text, let's merge this PR and come back to it. We can open an issue to add an equation to the docs.

> N.B. I think price can actually be negative

Interesting! I'm guessing that's not possible with the code as it stands though... (i.e. using the algorithm based on capacity duals)? 

> N.B. I feel like the parameter name cost_of_commodity is confusing. Can we change it to commodity_levy? Then document that a levy can be negative?

I agree "cost" is confusing. I've opened an issue to rename it everywhere: #553

  that the process could competitively serve assuming nothing changes from the previous
  milestone year).

  > For example, if demand is 8 units, and asset A serves 2 units
  of that at a marginal cost of $3, and asset B serves 4 units at
  a marginal cost of $5, but process C has a marginal cost of $4,
  then the algorithm assumes that process C could serve up to 6 units
  of demand in that time slice (though also limited by its

>   capacity/availability constraints in this and other time slices, etc).

Reviewer: I think this could be worded more clearly. I'll have a go at this later -- I'm just sticking a comment in now as a reminder

Author: OK

Reviewer: Actually, I think I misunderstood this the first time around. It makes more sense now.

  If the asset/process has lower marginal cost than any asset in a time slice, we assume
  dispatch only limited by capacity/availability constraints and demand level. If the
  process has the same marginal cost as an asset, we assume that asset is NOT displaced.
  If the process has marginal cost higher than any asset, we assume it can only serve demand
  currently unserved (i.e. due to asset decommissing at 3(A) or demadn increase), if any exists.
  If there is no asset (e.g. where demand was zero in previous milestone year, or all existing
  assets were decommissioned) then we assume dispatch only limited by process capacity/availability
  constraints and demand level.

> ##### ii. Calculate objective and decicion values for each asset/process

Reviewer: ```suggestion
##### ii. Calculate objective and decision values for each asset/process
```
